src/macos_native.rs

```diff
@@ -0,0 +1,892 @@
+mod ffi;
+use core_foundation::{
+    array::CFArray,
+    base::{kCFAllocatorDefault, CFGetTypeID, CFType, TCFType},
+    data::CFData,
+    dictionary::{CFDictionary, CFMutableDictionary},
+    mach_port::CFIndex,
+    number::CFNumber,
+    runloop::{
+        kCFRunLoopDefaultMode, CFRunLoop, CFRunLoopRunInMode, CFRunLoopRunResult, CFRunLoopSource,
+        CFRunLoopSourceContext, CFRunLoopSourceCreate, CFRunLoopSourceSignal, CFRunLoopStop,
+        CFRunLoopWakeUp,
+    },
+    string::CFString,
+};
+use mach2::port::MACH_PORT_NULL;
+
+use std::{
+    cell::Cell,
+    collections::VecDeque,
+    ffi::{c_void, CStr, CString},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Barrier, Condvar, Mutex,
+    },
+    time::Duration,
+};
+
+use crate::{
+    macos_native::ffi::{
+        kIOHIDProductIDKey, kIOHIDVendorIDKey, kIOReturnSuccess, IOHIDDevice,
+        IORegistryEntryFromPath,
+    },
+    DeviceInfo, HidDeviceBackendBase, HidDeviceBackendMacos, HidError, HidResult, WcharString,
+};
+
+use self::ffi::{io_registry_entry_get_registry_entry_id, kIOHIDDeviceUsagePairsKey, IOOptionBits};
+use self::ffi::{io_service_t, kIORegistryIterateParents};
+use self::ffi::{kIOHIDManufacturerKey, kIOHIDSerialNumberKey, IORegistryEntryIDMatching};
+use self::ffi::{kIOHIDProductKey, IORegistryEntrySearchCFProperty};
+use self::ffi::{kIOHIDReportType, kIORegistryIterateRecursively};
+use self::ffi::{kIOHIDTransportKey, IOServiceGetMatchingService};
+use self::ffi::{kIOHIDVersionNumberKey, IOReturn};
+use self::ffi::{kIOMainPortDefault, IOHIDManager};
+
+#[derive(Debug)]
+pub struct HidDevice {
+    /// If set to true, reads will block until data is available
+    blocking: Cell<bool>,
+
+    /// Options used to open the device
+    open_options: IOOptionBits,
+
+    /// Handle of thread responsible for reading from the device
+    reader_thread_handle: Option<std::thread::JoinHandle<()>>,
+
+    /// State shared beween reader thread and others
+    shared_state: Arc<SharedState>,
+}
+
+#[derive(Debug)]
+struct SharedState {
+    // Run loop mode used to read from the device
+    run_loop_mode: String,
+
+    max_input_report_len: usize,
+
+    // Reference to the run loop used to read from the device
+    run_loop: Mutex<Option<WrappedCFRunLoop>>,
+    source: Mutex<Option<LoopSource>>,
+
+    device: Mutex<IOHIDDevice>,
+
+    disconnected: AtomicBool,
+    shutdown_thread: AtomicBool,
+    shutdown_barrier: Barrier,
+
+    // Condition variable linked to input_reports
+    condition: std::sync::Condvar,
+    input_reports: Mutex<VecDeque<Vec<u8>>>,
```

Comment:
Use an `mpsc::channel` rather than a `VecDeque`. That way you get queue size limits for free, you don't need to wrap everything in a `Mutex`.

You also get helpers like `recv_timeout`...

Reply:
Right now, the oldest received input report is dropped when the limit is reached. I don't think this is possible with the `mpsc::sync_channel`, the size limit there could only be used stop sending new input reports, but it doesn't seem possible to remove elements from the channel.

Using a channel would be nice, but it would change the behaviour compared to the `C` hidapi library.